MNT Copy the CPython resource tracker code rather than derive from it

lesteve opened the pull request:
This may be preferable to avoid being affected by a CPython change in the future as in https://github.com/joblib/loky/issues/459, especially since now we override the `_ensure_running_and_write` method.

I tried to mostly add the CPython 3.13.7 [resource_tracker code](https://github.com/python/cpython/blob/v3.13.7/Lib/multiprocessing/resource_tracker.py) from the non-overriden methods and left things out where I was able to quickly see it was not used e.g. `self._reentrant_messages`.

We could align the code even more in a further PR, e.g. single vs double quotes (easy mostly needs to tell black to ignore this file) or some other things (reentrant logic, some small logic change in `main`, etc ...).

lesteve committed:
MNT Copy the CPython resource tracker code rather than derive from it

lesteve committed:
lint

lesteve committed:
lint with black

lesteve committed:
Adapt code to be compatible with different Python versions

lesteve committed:
Move __del__ up to better align with CPython 3.13 code + _reentrant_call_error inside PY311 block

lesteve committed:
Revert import ordering change

ogrisel commented:
There are teardown errors after `tests/test_reusable_executor.py::TestExecutorDeadLock::test_crash_races[13]` and subsequent tests on CPython 3.14.0rc2t.

Maybe we should directly sync with the code of the current `main` of CPython instead of CPython 3.13.7.

+1 for using the same code style as in CPython and configure the linter to ignore this file to ease future synchronizations.

frenzymadness commented:
Copying CPython code should lead to updating the license of this project or at least mentioning the Python license somewhere.

tomMoral commented:
As `loky` is mostly an extension/rewrite of CPython multiprocessing, we already have plenty of code that is copied from CPython so I would say this does not change much.

Note that according to Claude (our new law expert :sweat_smile: ):
> CPython is under the PSF License v2, which is a permissive BSD-style license and is compatible with BSD 3-Clause. So yes, BSD 3-Clause for your library is fine.

ogrisel commented:
We still need to acknowledge attribution. We should do so in the license file and the README.md.

lesteve committed:
Merge branch 'master' of https://github.com/joblib/loky into copy-python-stdlib-resource-tracker-rather-than-derive

lesteve commented:
CI is green (except codecov). I seem to remember there was some kind of problem with this PR, but apparently not or at least it's not a CI failure

In principle, I guess the additional decoding can be done for Python 3.15 as in https://github.com/joblib/loky/pull/626.

lesteve commented:
This is ready for review I think.

Here is a summary:
- there are some TODO left but I suggest Windows-specific behaviour and exit code handling can wait for a further PR. There are a few others that you may want to look at more closely.
- the generic approach is to minimize the diff compared to the stdlib code, which may add more diff compared to the previous loky version, sorry about this but I think this is the right call for easier maintainability in the future.
- Python 3.15 tests pass on my machine. This is a consequence of vendoring the stdlib code, we are not affected by CPython resource tracker changes (that sometimes can even happen in minor versions when fixes are back-ported).

cc @ogrisel (and @tomMoral in the off chance you miraculously get some unexpected spare bandwidth :wink:).

Also @larsoner, feel free to have a look if you want, since the faster we merge this, the sooner we can merge your Windows fixes https://github.com/joblib/loky/pull/643.

larsoner commented on loky/backend/resource_tracker.py:
```diff
@@ -345,60 +413,96 @@ def main(fd, verbose=0):
+        # is called for all resources except folders and then at the end for
+        # all folders
+        def _unlink_resources(rtype_cache, rtype):
+            for rtype, rtype_cache in cache.items():
```
These two vars shadow the inputs, this looks like a bug? If it's intentional then at least having different names here would make sense. But I'm guessing the `for` should be gone here and lines dedented?

larsoner commented on loky/backend/stdlib_py314_resource_tracker.py:
```diff
@@ -0,0 +1,495 @@
+        if self._use_simple_format:
+            return b'PROBE:0:noop\n'
+        return (
+            json.dumps(
```
I think `import json` is missing at the top...

... and that a good linter probably would have caught this. Happy to make a quick PR to use `ruff` like in `sklearn` at some point

larsoner commented on loky/backend/resource_tracker.py:
```diff
@@ -253,9 +284,41 @@ def __del__(self):
+            raise ValueError("malformed resource_tracker fields: %r" % (obj,))
+
+        try:
+            name = base64.urlsafe_b64decode(b64).decode('utf-8', 'surrogateescape')
```
`base64` also not imported

larsoner commented on loky/backend/resource_tracker.py:
```diff
@@ -44,9 +44,20 @@
+    )
+else:
+    # CPython >= 3.11 re-entrancy support code relies on
+    # threading.RLock._recursion_count which does not exist for Python <= 3.10
```
It also doesn't exist for some versions of 3.11 (<= 3.11.5) and 3.12 (<= 3.12.0) because it was a backport. A simpler possible check would be `hasattr(threading.RLock(), "_recursion_count")` (but maybe that's fragile?)

larsoner commented on pyproject.toml:
```diff
@@ -54,7 +54,7 @@ version = {attr = "loky.__version__"}
 include = '\.pyi?$'
 exclude = '''
-/(
+(
```
Intentional to drop repo-root leading-anchor `/` here? This will now skip stuff like `poky/backend/redistribute_stuff.py` because it contains the string `dist`

larsoner reviewed (commented):
Okay this is new stuff for me, so I had Opus review, and I then looked at some of the stuff it flagged, and I agree a few of them look like true oversights so I've commented inline on those. (The import errors suggest also that some tests could maybe be improved to hit the problematic paths maybe... and more generally updated to make sure the loky-specific paths are hit, but this is a bit more work still and probably not a blocker.) But I am new to this and Opus gets stuff wrong (even though I cross-checked its claims) so take it all with a grain of salt!

Then with my own brain 🧠 I had the thought: with the tag for #624, would it be worth trying to add a Python 3.15 CI run? You could do it with setup-python, but something like this should work for conda-forge (I had Opus draft these changes):

In `continuous_integration/install.sh`:
```diff
-to_install="$PYTHON_PACKAGE=$PYTHON_VERSION pip numpy tblib $EXTRA_PACKAGES"
-conda create -n testenv --yes -c $CONDA_CHANNEL $to_install
+to_install="$PYTHON_PACKAGE=$PYTHON_VERSION pip tblib $EXTRA_PACKAGES"
+# conda-forge has no numpy build for pre-release CPython; get it from PyPI instead
+if [[ "$NUMPY_FROM_PIP" != "true" ]]; then
+    to_install="$to_install numpy"
+fi
+conda create -n testenv --yes -c $CONDA_CHANNEL \
+    ${EXTRA_CONDA_CHANNEL:+-c $EXTRA_CONDA_CHANNEL} -c conda-forge $to_install

 if [[ -z "$NO_PSUTIL" ]]; then
     PIP_INSTALL_PACKAGES="$PIP_INSTALL_PACKAGES psutil"
 fi
+if [[ "$NUMPY_FROM_PIP" == "true" ]]; then
+    PIP_INSTALL_PACKAGES="$PIP_INSTALL_PACKAGES numpy"
+fi
```
And in the CI file
```
          - name: linux-py315
            os: ubuntu-latest
            PYTHON_VERSION: "3.15"
            CONDA_CHANNEL: "conda-forge/label/python_dev"
            EXTRA_CONDA_CHANNEL: "conda-forge/label/python_rc"
            NUMPY_FROM_PIP: "true"
            NO_PSUTIL: "true"   # psutil ships sdist only for cp315
```
I'm happy to try this conda env and iterate with Opus locally and open a PR to your branch (making sure the diff is sane) with these changes if it would help!

lesteve commented:
I fixed the stupid mistakes, thanks @larsoner!

OK my last attempt is to try to only have one stdlib vendored implementation i.e. the Python 3.14.7 one and have a switch depending on the lock `_recursion_count` thing. It may be a bit bold, but this is simpler and the hacky part is only for `Python 3.10.x` (whose support we can drop support in end October 2026), `Python>=3.11.0<3.11.6` and `Python 3.12.0` so that's acceptable?

The Windows Python 3.10 seems genuine, I need to have a closer look ...

A PR switching to `ruff` would be welcome at one point, but then maybe after this PR is merged to avoid having to solve conflicts? I guess I would need to figure out the equivalent tweaks for `ruff` to leave some parts of the file alone. A LLM is telling me you need something like this:
```py
  # fmt: off
  # ruff: disable[E501, F841]

  # code...

  # ruff: enable[E501, F841]
  # fmt: on
```

Same thing for adding Python 3.15 CI, once that PR is merged, this would be more than welcome :pray:! FWIW I am running locally the tests with Python 3.15 and they seem to pass.

larsoner commented:
Windows failure is:
```
Exception ignored in: <function ResourceTracker.__del__>
  File "...stdlib_py314_resource_tracker.py", line 156, in _stop_locked
AttributeError: '_thread.RLock' object has no attribute '_recursion_count'
```
Opus says d10f85b is to blame, and I buy it. The vendored code has:
```
if hasattr(os, 'register_at_fork'):
    os.register_at_fork(after_in_child=_resource_tracker._after_fork_in_child)
```
`os.register_at_fork` exists on POSIX, so the `_resource_tracker` lives forever. It doesn't exist on Windows, so the method is collected at interpreter shutdown, triggering the error via `ResourceTracker.__del__` -> `_stop` -> `_stop_locked` which then has the `if self._lock._recursion_count() > 1` on line 156.

One option would be strip the module-level block in `vendor_stdlib_resource_tracker.sh`. Another would be to shim it:
```
from . import stdlib_py314_resource_tracker as _stdlib_rt

if not hasattr(_stdlib_rt._resource_tracker._lock, "_recursion_count"):
    _stdlib_rt._resource_tracker._lock = LokyRLock()
```

The LLM also suggested that since `_thread.RLock` has `Py_TPFLAGS_BASETYPE` on 3.10.20 / 3.11.0 / 3.12.0, so class `LokyRLock(type(threading.RLock()))` works and keeps the C lock rather than the Python one (and it said "`threading._RLock` is ~3.2x slower per acquire/release, and this path is live for all of 3.10").

ogrisel commented on loky/backend/stdlib_py314_resource_tracker.py:
```diff
@@ -0,0 +1,495 @@
+
+# # gh-146313: See _after_fork_in_child docstring.
+# if hasattr(os, 'register_at_fork'):
+#     os.register_at_fork(after_in_child=_resource_tracker._after_fork_in_child)
```
Re-running the vendoring script removes those comments, hence the need for a patch. The patch could include extra inline comments to justify the changes.

ogrisel commented on tools/vendor_stdlib_resource_tracker.sh:
```diff
@@ -0,0 +1,7 @@
+     https://raw.githubusercontent.com/python/cpython/refs/tags/v3.14.7/Lib/multiprocessing/resource_tracker.py \
+     -o loky/backend/stdlib_py314_resource_tracker.py
+
+sed -i 's@from \. import@from multiprocessing import@' loky/backend/stdlib_py*_resource_tracker.py
```
To make it also work with BSD sed (e.g. on macOS), we need something like. But as discussed IRL, maintaining a real patch might be a better idea.

```suggestion
sed -i.backup 's@from \. import@from multiprocessing import@' loky/backend/stdlib_py*_resource_tracker.py
```

ogrisel reviewed (commented):
Quick feedback not too forgot. Feel free to ping me once ready for review.

cakedev0 commented on loky/backend/resource_tracker.py:
```diff
@@ -253,9 +236,41 @@ def __del__(self):
 
+# Copied from Python 3.14.7
+# fmt: off
+def _decode_message(line):
```
We could import this from the vendored code (exact same code)

cakedev0 commented on loky/backend/resource_tracker.py:
```diff
@@ -106,144 +124,109 @@ class ResourceTracker(_ResourceTracker):
+                self._stop(use_blocking_lock=False)
+            except ChildProcessError:
+                # ignore error due to trying to clean up child process which has already been
+                # shutdown on windows. See https://github.com/joblib/loky/pull/450
```
You can try that to fix the windows 3.10 CI :crossed_fingers: 

```suggestion
            except OSError:
                # ignore error due to trying to clean up child process which has already been
                # shutdown on windows. See https://github.com/joblib/loky/pull/450
                # ChildProcessError is a subclass of OSError, but other OSError
                # variants can also be raised by os.waitpid on Windows during
                # interpreter shutdown; this is best-effort cleanup either way.
```

lesteve replied:
I think this is no longer relevant, since our approach is to override the _stop function on Windows.

cakedev0 commented on loky/backend/resource_tracker.py:
```diff
@@ -55,43 +59,57 @@
+# TODO Remove work-around when Python 3.13 is our minimum supported version
+class LokyRLock(_RLock):
+    def _recursion_count(self):
+        return 1
```
_owner and _count are defined for the versions of Python where this lock will be used. So I think we ca do this:

```suggestion
        if self._owner != get_ident():
            return 0
        return self._count
```

cakedev0 replied:
(needs `from threading import _RLock, get_ident` on top)

lesteve replied:
That doesn't seem to work, maybe when we inherit from the C lock (rather than the Python lock)? For example on Python 3.10

```
❯ pytest tests/test_resource_tracker.py -x 
===================================================================================================================================================== test session starts ======================================================================================================================================================
platform linux -- Python 3.10.20, pytest-9.1.1, pluggy-1.6.0
rootdir: /home/lesteve/dev/loky
configfile: pyproject.toml
collected 16 items                                                                                                                                                                                                                                                                                                             

tests/test_resource_tracker.py ...F

=========================================================================================================================================================== FAILURES ===========================================================================================================================================================
__________________________________________________________________________________________________________________________________ TestResourceTracker.test_child_retrieves_resource_tracker ___________________________________________________________________________________________________________________________________

self = <tests.test_resource_tracker.TestResourceTracker object at 0x7fd20b0f44f0>

    def test_child_retrieves_resource_tracker(self):
    
        # First simple fd retrieval check (see #200)
        # checking fd only work on posix for now
        if sys.platform != "win32":
            try:
>               parent_rtracker_fd = get_rtracker_fd()

tests/test_resource_tracker.py:50: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_resource_tracker.py:29: in get_rtracker_fd
    resource_tracker.ensure_running()
loky/backend/stdlib_py314_resource_tracker.py:215: in ensure_running
    return self._ensure_running_and_write()
loky/backend/stdlib_py314_resource_tracker.py:291: in _ensure_running_and_write
    if self._lock._recursion_count() > 1:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <unlocked LokyRLock object owner=0 count=0 at 0x7fd20b0d2b60>

    def _recursion_count(self):
>       if self._owner != get_ident():
E       AttributeError: 'LokyRLock' object has no attribute '_owner'

loky/backend/resource_tracker.py:113: AttributeError

During handling of the above exception, another exception occurred:

self = <tests.test_resource_tracker.TestResourceTracker object at 0x7fd20b0f44f0>

    def test_child_retrieves_resource_tracker(self):
    
        # First simple fd retrieval check (see #200)
        # checking fd only work on posix for now
        if sys.platform != "win32":
            try:
                parent_rtracker_fd = get_rtracker_fd()
                executor = ProcessPoolExecutor(max_workers=2)
                child_rtracker_fd = executor.submit(get_rtracker_fd).result()
    
                assert child_rtracker_fd == parent_rtracker_fd
            finally:
>               executor.shutdown()
E               UnboundLocalError: local variable 'executor' referenced before assignment

tests/test_resource_tracker.py:56: UnboundLocalError
=================================================================================================================================================== short test summary info ====================================================================================================================================================
FAILED tests/test_resource_tracker.py::TestResourceTracker::test_child_retrieves_resource_tracker - UnboundLocalError: local variable 'executor' referenced before assignment
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================================================================================================= 1 failed, 3 passed in 0.12s ==================================================================================================================================================
[INFO:MainProcess:MainThread] process shutting down
[DEBUG:MainProcess:MainThread] running all "atexit" finalizers with priority >= 0
[DEBUG:MainProcess:MainThread] running the remaining "atexit" finalizers
```

cakedev0 commented on loky/backend/resource_tracker.py:
```diff
@@ -253,9 +236,41 @@ def __del__(self):
 unregister = _resource_tracker.unregister
 getfd = _resource_tracker.getfd
 
```
You could add here this piece that is commented out in the vendored code. Apparently it guards against some edge cases (also, if you don't do that, the vendored code `_after_fork_in_child` method is dead code).

```suggestion

# gh-146313: See _after_fork_in_child docstring.
if hasattr(os, 'register_at_fork'):
    os.register_at_fork(after_in_child=_resource_tracker._after_fork_in_child)

```

cakedev0 reviewed (commented):
Some findings from AI, hopefully useful/worth fixing.

Once address, I can approve this PR even if I don't feel super legitimate as my understanding is still quite shallow. But the overall direction (vendoring for stability) feels right, and the tests mostly pass ^^

lesteve commented:
I incorporated the fixes from #642 in this PR in https://github.com/joblib/loky/pull/472/commits/2813ad8965fe6884efb8d0cdaef271d39db07b53 and added you as a co-author @larsoner, hopefully this is fine with you?

Basically I didn't want to have to figure out whether the remaining Windows failures were my fault or not. CI is green the first time. I'll push a few commits to check it wasn't a lucky break. Edit: CI is green 5 times out of 5 [^1] we can confidently say that the Windows issue has been fixed by incorporating the fixes from #642.

[^1]: codecov is red but all the rest is green

larsoner commented:
Yes that's fine!

ogrisel commented on loky/backend/resource_tracker.py:
Let's document how this file is supposed to be maintained.


```suggestion
#
# Source code organization and maintenance:
#
# This file is derived from the matching file in the standard library with
# modifications to add the new features described above.
#
# It defines a subclass of a vendored copy of the ResourceTracker from the
# standard library to make loky less likely to break when internals of
# change in a new version of CPython.
#
# The new features are implemented in the overriden method of the ResourceTracker
# class as well as in the custom `main` function.
#
# The vendored copy is minimally patched and should not be edited by hand.
# Instead, it should be revendored from time to time using the script
# in the `tools/` folder at the root of this repo.
#
# When vendoring a new copy of the resource_tracker module of the standard
# library, it is important to check of the `main` function has evolved
# upstream. If so, doing a diff of that particular function in the upstream
# file and in the current file might be helpful.
#
# Make sure to update the inline comments to help future maintainers understand
# what loky-specific changes where made.
```

ogrisel replied:
Maybe we could even provide a helper script to diff the main function. Here is a proposal by cursor:

```
python3 - <<'PY' file_a.py file_b.py main
import ast, sys
from pathlib import Path

def extract(path, name):
    src = Path(path).read_text()
    tree = ast.parse(src)
    for node in tree.body:
        if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)) and node.name == name:
            return ast.get_source_segment(src, node)
    raise SystemExit(f"{name!r} not found in {path}")

a, b, name = sys.argv[1], sys.argv[2], sys.argv[3]
Path("/tmp/_fa.py").write_text(extract(a, name) + "\n")
Path("/tmp/_fb.py").write_text(extract(b, name) + "\n")
PY
diff -u /tmp/_fa.py /tmp/_fb.py
```

ogrisel reviewed (commented):
I had a look at the diff of the PR and LGTM. However, I think we could improve the tooling for future vendoring/updates.

If we keep the sed based solution, we should use a suffix for -i flag to make it work with BSD. Alternatively we could use a real patch.

See also:

ogrisel commented on loky/backend/resource_tracker.py:
```diff
@@ -36,16 +36,44 @@
+# file and in the current file might be helpful.
+#
+# Make sure to update the inline comments to help future maintainers understand
+# what loky-specific changes where made.
```
```suggestion
# what loky-specific changes were made.
```

ogrisel commented on loky/backend/resource_tracker.py:
```diff
@@ -55,43 +83,62 @@
 
+# loky: Windows support
+# Windows process waits use milliseconds.
+# TODO now: only used in tests???
```
Indeed, let's move to the test file instead.

ogrisel commented on loky/backend/resource_tracker.py:
```diff
@@ -345,67 +408,111 @@ def main(fd, verbose=0):
+                            util.debug(f'[ResourceTracker] unlink {name}')
+                    except Exception as e:
+                        exit_code = 2
+                        # loky: tweaked formatting ($r instead of %s for
```
```suggestion
                        # loky: tweaked formatting (%r instead of %s for
```

lesteve commented on loky/backend/resource_tracker.py:
```diff
@@ -345,67 +403,111 @@ def main(fd, verbose=0):
+        # it, maybe Windows ... git blame points at
+        # https://github.com/joblib/loky/pull/429 but couldn't find any clear
+        # reason
         args = [arg.encode("utf-8") for arg in args]
```
@ogrisel any chance you remember the reason for the `arg.encode` here? stdlib doesn't have it and it was done in https://github.com/joblib/loky/pull/429 and I could not find a clear reason in the PR discussion ...

I guess if we don't have a good reason, we might as well remove it ...

lesteve replied:
Let's remove it, a LLM is telling me that there is no strong reason and it may be related to the fact that we were using `fork_exec` before where explicit encoding was relevant

ogrisel replied:
Maybe we could try confirming if this is needed or not by running some unit tests with non-ascii arguments or non ascii characters in the folder names in the path?

ogrisel replied:
Note that on Windows, the default system encoding is typically not UTF-8 (even in recent versions, contrary to Linux and macOS). So if the system encoding used by default cannot represent a particular unicode symbol we might get an encoding error.

ogrisel replied:
For instance, we could try to write a test with ✅ (U+2705, WHITE HEAVY CHECK MARK) as a filename or argument. On Linux `touch ✅` and `ls -l ✅` work. I am not sure if it would work on Windows with a typical default locale setting.

lesteve commented:
Oh well a weird error on macOS that I am going to ignore unless it happens again ... see [build log](https://github.com/joblib/loky/actions/runs/32379589482/job/96459255921?pr=472)

```
_______________ TestsProcessPoolLokyExecutor.test_worker_timeout _______________
PermissionError: [Errno 13] force permission denied (originated from sysctl(KERN_PROCARGS2) -> errno 0)
```


<details>
<summary>Details</summary>

```
_______________ TestsProcessPoolLokyExecutor.test_worker_timeout _______________
PermissionError: [Errno 13] force permission denied (originated from sysctl(KERN_PROCARGS2) -> errno 0)

The above exception was the direct cause of the following exception:

self = <tests.test_process_executor_loky.TestsProcessPoolLokyExecutor object at 0x1085701f0>

    @pytest.mark.timeout(60)
    def test_worker_timeout(self):
        self.executor.shutdown(wait=True)
        self.check_no_running_workers(patience=5)
        timeout = getattr(self, "min_worker_timeout", 0.01)
        try:
            self.executor = self.executor_type(
                max_workers=4, context=self.context, timeout=timeout
            )
        except NotImplementedError as e:
            self.skipTest(str(e))
    
        for _ in range(5):
            with warnings.catch_warnings():
                # It's ok to get a warning about the worker interrupted by the
                # short timeout while tasks are pending in the queue on
                # overloaded CI hosts.
                warnings.simplefilter("ignore", category=UserWarning)
    
                # Trigger worker spawn for lazy executor implementations
                for _ in self.executor.map(id, range(8)):
                    pass
    
            # Check that all workers shutdown (via timeout) when waiting a bit:
            # note that the effective time for a Python process to completely
            # shutdown can vary a lot especially on loaded CI machines with and
            # the atexit callbacks that writes test coverage data to disk.
            # Let's be patient.
>           self.check_no_running_workers(patience=5)

_          = 4369656360
self       = <tests.test_process_executor_loky.TestsProcessPoolLokyExecutor object at 0x1085701f0>
timeout    = 0.01

tests/_test_process_executor.py:902: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/_executor_mixin.py:246: in check_no_running_workers
    workers = _running_children_pids_with_cmdline(p)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        cls        = <class 'tests.test_process_executor_loky.TestsProcessPoolLokyExecutor'>
        deadline   = 1787235836.244681
        p          = psutil.Process(pid=2454, name='python3.13', status='running')
        patience   = 5
        sleep_duration = 0.01
        workers    = [(4954, '/Users/runner/miniconda3/envs/testenv/bin/python3.13 -m loky.backend.popen_loky_posix --process-name LokyProc...ner/miniconda3/envs/testenv/bin/python3.13 -m loky.backend.popen_loky_posix --process-name LokyProcess-608 --pipe 81')]
tests/_executor_mixin.py:60: in _running_children_pids_with_cmdline
    all_children = _direct_children_with_cmdline(p)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        p          = psutil.Process(pid=2454, name='python3.13', status='running')
tests/_executor_mixin.py:43: in _direct_children_with_cmdline
    cmdline = " ".join(c.cmdline())
                       ^^^^^^^^^^^
        c          = psutil.Process(pid=4955, status='terminated', started='14:23:51')
        children_with_cmdline = [(psutil.Process(pid=2508, name='python3.13', status='running', started='14:22:02'), '/Users/runner/miniconda3/envs/te...er/miniconda3/envs/testenv/bin/pytest -vl --timeout=120 --maxfail=5 --cov=loky --cov-report xml --skip-high-memory .')]
        cmdline    = '/Users/runner/miniconda3/envs/testenv/bin/python3.13 -c import sys; from multiprocessing.forkserver import main; main...nner/miniconda3/envs/testenv/bin/pytest -vl --timeout=120 --maxfail=5 --cov=loky --cov-report xml --skip-high-memory .'
        p          = psutil.Process(pid=2454, name='python3.13', status='running')
../../../miniconda3/envs/testenv/lib/python3.13/site-packages/psutil/__init__.py:748: in cmdline
    return self._proc.cmdline()
           ^^^^^^^^^^^^^^^^^^^^
        self       = psutil.Process(pid=4955, status='terminated', started='14:23:51')
../../../miniconda3/envs/testenv/lib/python3.13/site-packages/psutil/_psosx.py:347: in wrapper
    return fun(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
        args       = ()
        fun        = <function Process.cmdline at 0x1077f0ae0>
        kwargs     = {}
        name       = None
        pid        = 4955
        ppid       = None
        self       = <psutil._psosx.Process object at 0x108676680>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <psutil._psosx.Process object at 0x108676680>

    @wrap_exceptions
    def cmdline(self):
>       return cext.proc_cmdline(self.pid)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       SystemError: <built-in function proc_cmdline> returned a result with an exception set

self       = <psutil._psosx.Process object at 0x108676680>

../../../miniconda3/envs/testenv/lib/python3.13/site-packages/psutil/_psosx.py:405: SystemError
```

</details>

larsoner commented:
If you want to fix that flake, Claude suggested:

---

One line in tests/_executor_mixin.py:
```
-        except (OSError,) + psutil_exceptions:
+        # SystemError: psutil <= 7.2.2 on macOS returns a result with a
+        # pending exception when the process exits mid-call (psutil #2854)
+        except (OSError, SystemError) + psutil_exceptions:
```

ogrisel commented on tools/README.md:
```diff
@@ -0,0 +1,43 @@
+from a fixed CPython version. Loky minimal changes to that file should be
+stored as a patch in `tools/stdlib_py314_resource_tracker.patch`. As a patch, it
+has more chances to still apply if the CPython code change and we want to
+update the vendored stdlib copy.
```
```suggestion
`vendor_stdlib_resource_tracker.sh` downloads the `resource_tracker.py` file
from a fixed CPython version. Loky-specific minimal changes to that file should
be stored as a patch in `tools/stdlib_py314_resource_tracker.patch`.

As a minimal patch, it has more chances to still apply when re-vendorig a newer
versions of the CPython resource tracker module with some code changes, e.g. to
benefit from upstream bug fixes.
```

ogrisel commented on tools/README.md:
```diff
@@ -0,0 +1,43 @@
+- modify `loky/backend/stdlib_py314_resource_tracker.py` if really needed
+  (remember changes should be minimal to avoid diverging from the reference
+  CPython code)
+- Download the Python reference file
```
```suggestion
- download the Python reference file:
```

ogrisel commented on tools/README.md:
````diff
@@ -0,0 +1,43 @@
+      https://raw.githubusercontent.com/python/cpython/refs/tags/v3.14.7/Lib/multiprocessing/resource_tracker.py \
+      -o loky/backend/stdlib_py314_resource_tracker.py.ref
+  ```
+- generate the patch
````
```suggestion
- generate the patch:
```

ogrisel commented on tools/README.md:
````diff
@@ -0,0 +1,43 @@
+      https://raw.githubusercontent.com/python/cpython/refs/tags/v3.14.7/Lib/multiprocessing/resource_tracker.py \
+      -o loky/backend/stdlib_py314_resource_tracker.py.ref
+  ```
+- generate the patch
````
```suggestion
- delete the old patch: `rm -rf tools/stdlib_py314_resource_tracker.patch`
- generate the updated patch:
```

lesteve replied:
I tweaked the command to overwrite the file. TIL `>|` works on bash and zsh, I knew `>!` was zsh-specific.

ogrisel reviewed (approved):
I followed the instructions to vendor and refresh the patch and it works well besides the following nits.

Note that when refreshing the patch, I get the following local changes:

```diff
diff --git a/tools/stdlib_py314_resource_tracker.patch b/tools/stdlib_py314_resource_tracker.patch
index 81378c4..44ea729 100644
--- a/tools/stdlib_py314_resource_tracker.patch
+++ b/tools/stdlib_py314_resource_tracker.patch
@@ -1,5 +1,5 @@
---- loky/backend/stdlib_py314_resource_tracker.py.ref  2026-08-20 12:38:59.671027411 +0200
-+++ loky/backend/stdlib_py314_resource_tracker.py      2026-08-20 12:38:47.394820500 +0200
+--- loky/backend/stdlib_py314_resource_tracker.py.ref  2026-08-21 12:05:32
++++ loky/backend/stdlib_py314_resource_tracker.py      2026-08-21 12:08:32
 @@ -26,8 +26,9 @@
  
  import json
@@ -16,26 +16,25 @@
  # popen_fork.Popen._launch() at once from clobbering eachothers intent.
  _fork_intent = threading.local()
  
-+# loky: begin commented out section
-+# Remove module-level globals and logic which
-+# are unused and can cause confusing errors at interpreter exit because of
-+# discrepancies in the code from vendored Python version and used Python
-+# version
 -_resource_tracker = ResourceTracker()
 -ensure_running = _resource_tracker.ensure_running
 -register = _resource_tracker.register
 -unregister = _resource_tracker.unregister
 -getfd = _resource_tracker.getfd
--
--# gh-146313: See _after_fork_in_child docstring.
--if hasattr(os, 'register_at_fork'):
--    os.register_at_fork(after_in_child=_resource_tracker._after_fork_in_child)
++# loky: begin commented out section
++# Remove module-level globals and logic which
++# are unused and can cause confusing errors at interpreter exit because of
++# discrepancies in the code from vendored Python version and used Python
++# version
 +# _resource_tracker = ResourceTracker()
 +# ensure_running = _resource_tracker.ensure_running
 +# register = _resource_tracker.register
 +# unregister = _resource_tracker.unregister
 +# getfd = _resource_tracker.getfd
-+
+ 
+-# gh-146313: See _after_fork_in_child docstring.
+-if hasattr(os, 'register_at_fork'):
+-    os.register_at_fork(after_in_child=_resource_tracker._after_fork_in_child)
 +# # gh-146313: See _after_fork_in_child docstring.
 +# if hasattr(os, 'register_at_fork'):
 +#     os.register_at_fork(after_in_child=_resource_tracker._after_fork_in_child)
```

lesteve commented:
> Note that when refreshing the patch, I get the following local changes:

I tweaked the command to remove the diff timestamps in my last commit. Other than this, I have seen a similar thing locally and it seems both patches are equivalent which a LLM is telling me can happen (`diff` is not guaranteed to produce a unique result). I added `--mimimal` which may help ...

Other than this, trying to make sense of a diff of a diff (second-order derivative) makes me :exploding_head: :sweat_smile:.

ogrisel commented on loky/backend/resource_tracker.py:
```diff
@@ -106,156 +148,172 @@ class ResourceTracker(_ResourceTracker):
-            "Some folders/semaphores might leak."
-        )
+            super()._teardown_dead_process()
+        else:
```
It would be cleaner to condition explicitly for win32 in an `elif` instead.


```suggestion
        elif sys.platform == "win32":
```

lesteve commented:
> Merge?

We have never been closer I think :sweat_smile:. `/review` tells me this on Codex. The second one makes sense, the first one I need to look at it closer to actually understand what it means ...

• The at-fork registration is not paired with the multiprocessing fork-intent integration, which can disconnect forked children from the inherited tracker. Cleanup status assignments also fail to affect the process exit code.

  Full review comments:

  - [P1] Preserve the tracker descriptor in multiprocessing fork children — /home/lesteve/dev/loky/loky/backend/resource_tracker.py:311-312
    When a process is created with multiprocessing's fork start method on Python versions whose popen_fork does not set this vendored module's _fork_intent, this callback treats the fork as a raw os.fork() and closes the inherited loky tracker descriptor. If the child subsequently calls maybe_unlink or otherwise uses
    an inherited tracked resource, it launches an empty tracker instead of updating the parent's refcount, causing protocol errors or premature/leaked cleanup. The matching popen_fork integration must be provided, or this callback should not be registered.

  - [P2] Propagate cleanup failures to the tracker exit code — /home/lesteve/dev/loky/loky/backend/resource_tracker.py:435-438
    When leaked resources are found or cleanup fails, these assignments create a local exit_code inside _unlink_resources because the nested function has no nonlocal exit_code. Consequently main() still exits with 0 unless message processing previously failed, and _resource_tracker._exitcode cannot report the newly
    intended leak/cleanup statuses. Declare the variable nonlocal or return and combine the cleanup status.

ogrisel commented:
I think it would help if you could generate some reproducer test code for the two problems before attempting to fix them.

ogrisel commented on tests/test_resource_tracker.py:
```diff
@@ -289,6 +338,123 @@ def test_resource_tracker_sigkill(self):
+        not hasattr(os, "register_at_fork"),
+        reason="os.register_at_fork unavailable",
+    )
+    def test_resource_tracker_at_fork_callback_is_registered(self):
```
```suggestion
    def test_resource_tracker_at_fork_callback_is_registered(self):
        # Ensure that `_resource_tracker._after_fork_in_child` is called to
        # perform the fd handling specified when the `_fork_intent` module-level
        # variable from the multiprocessing module of the stdlib (rather than
        # the loky vendored copy) is set to False.
        #
        # After a raw os.fork(), the child's ResourceTracker._pid must be None.
        # That only happens if os.register_at_fork registered
        # _resource_tracker._after_fork_in_child
```

ogrisel commented on tests/test_resource_tracker.py:
```diff
@@ -289,6 +338,123 @@ def test_resource_tracker_sigkill(self):
+        "fork" not in multiprocessing.get_all_start_methods(),
+        reason="fork start method unavailable",
+    )
+    def test_multiprocessing_fork_preserves_resource_tracker_fd(self):
```
```suggestion
    def test_multiprocessing_fork_preserves_resource_tracker_fd(self):
        # Check that multiprocessing's `preserve_fd=True` is respected.
```

ogrisel commented on loky/backend/resource_tracker.py:
```diff
@@ -345,67 +416,106 @@ def main(fd, verbose=0):
+                            util.debug(f'[ResourceTracker] unlink {name}')
+                    except Exception as e:
+                        exit_code = 2
+                        # loky: tweaked formatting ($r instead of %s for
```
```suggestion
                        # loky: tweaked formatting (%r instead of %s for
```

ogrisel commented on tests/test_resource_tracker.py:
```diff
@@ -289,6 +338,123 @@ def test_resource_tracker_sigkill(self):
+    ):
+        winapi = _RecordingWinapi()
+        monkeypatch.setattr(resource_tracker, "_winapi", winapi, raising=False)
+        monkeypatch.setattr(os, "name", "nt")
```
```suggestion
        monkeypatch.setattr(os, "name", "nt")
        monkeypatch.setattr(sys, "platform", "win32")
```

lesteve commented:
The last two problems from https://github.com/joblib/loky/pull/472#issuecomment-5393087133 have been fixed and tests have been added.

All right, let's merge this! CI has been consistently green in all the "good" pushes [^1], which is nice :tada:!

Thanks everyone for the feed-back :pray:!

[^1]: commit looks red only because of codecov all the other builds are green. There were also some "bad" pushes of course :sweat_smile:
